File: java/08_007.java
import java.util.ArrayDeque;
import java.util.Deque;

public class SlidingWindowExtremaProduct {
    public int[] findExtremaProducts(int[] prices, int k) {
        int n = prices.length;
        
        if (n == 0 || k == 0 || k > n) {
            return new int[0];
        }
        
        int[] result = new int[n - k + 1];
        Deque<Integer> maxDeque = new ArrayDeque<>();
        Deque<Integer> minDeque = new ArrayDeque<>();
        
        // Process the first window
        for (int i = 0; i < k; i++) {
            // Remove smaller elements from maxDeque
            while (!maxDeque.isEmpty() && 
                   prices[i] >= prices[maxDeque.peekLast()]) {
                maxDeque.pollLast();
            }
            maxDeque.offerLast(i);
            
            // Remove larger elements from minDeque
            while (!minDeque.isEmpty() && 
                   prices[i] <= prices[minDeque.peekLast()]) {
                minDeque.pollLast();
            }
            minDeque.offerLast(i);
        }
        
        // Calculate product for the first window
        result[0] = prices[maxDeque.peekFirst()] * 
                    prices[minDeque.peekFirst()];
        
        // Process the rest of the windows
        for (int i = k; i < n; i++) {
            // Remove elements outside the current window
            while (!maxDeque.isEmpty() && 
                   maxDeque.peekFirst() <= i - k) {
                maxDeque.pollFirst();
            }
            while (!minDeque.isEmpty() && 
                   minDeque.peekFirst() <= i - k) {
                minDeque.pollFirst();
            }
            
            // Remove smaller elements from maxDeque
            while (!maxDeque.isEmpty() && 
                   prices[i] >= prices[maxDeque.peekLast()]) {
                maxDeque.pollLast();
            }
            maxDeque.offerLast(i);
            
            // Remove larger elements from minDeque
            while (!minDeque.isEmpty() && 
                   prices[i] <= prices[minDeque.peekLast()]) {
                minDeque.pollLast();
            }
            minDeque.offerLast(i);
            
            // Calculate product for the current window
            result[i - k + 1] = prices[maxDeque.peekFirst()] * 
                               prices[minDeque.peekFirst()];
        }
        
        return result;
    }
}

File: python/08_007.py
from collections import deque

def sliding_window_extrema_product(prices, k):
    n = len(prices)
    result = []
    
    if n == 0 or k == 0 or k > n:
        return result
    
    max_deque = deque()
    min_deque = deque()
    
    # Process the first window
    for i in range(k):
        # Remove smaller elements from max_deque
        while max_deque and prices[i] >= prices[max_deque[-1]]:
            max_deque.pop()
        max_deque.append(i)
        
        # Remove larger elements from min_deque
        while min_deque and prices[i] <= prices[min_deque[-1]]:
            min_deque.pop()
        min_deque.append(i)
    
    # Calculate product for the first window
    result.append(prices[max_deque[0]] * prices[min_deque[0]])
    
    # Process the rest of the windows
    for i in range(k, n):
        # Remove elements outside the current window
        while max_deque and max_deque[0] <= i - k:
            max_deque.popleft()
        while min_deque and min_deque[0] <= i - k:
            min_deque.popleft()
        
        # Remove smaller elements from max_deque
        while max_deque and prices[i] >= prices[max_deque[-1]]:
            max_deque.pop()
        max_deque.append(i)
        
        # Remove larger elements from min_deque
        while min_deque and prices[i] <= prices[min_deque[-1]]:
            min_deque.pop()
        min_deque.append(i)
        
        # Calculate product for the current window
        result.append(prices[max_deque[0]] * 
                     prices[min_deque[0]])
    
    return result